graphistry/plugins/spannergraph.py

```diff
@@ -0,0 +1,294 @@
+            return connection
+        except Exception as e:
+            raise SpannerConnectionError(f"Failed to connect to Spanner: {e}")
+
+    def close_connection(self) -> None:
+        """
+        Closes the connection to the Spanner database.
+        """
+        if self.connection:
+            self.connection.close()
+            logging.info("Connection to Spanner database closed.")
+
+    def execute_query(self, query: str) -> SpannerQueryResult:
+        """
+        Executes a GQL query on the Spanner database.
+
+        :param query: The GQL query to execute.
+        :return: The results of the query execution.
+        :rtype: SpannerQueryResult
+        :raises RuntimeError: If the query execution fails.
+        """
+        logger.debug(f' SpannerGraph execute_query() query:{query}\n')
+
+        try:
+            start_time = time.time()
+            cursor = self.connection.cursor()
+            cursor.execute(query)
+            results = cursor.fetchall()
+            execution_time = time.time() - start_time
+            logging.info(f"Query executed in {execution_time:.4f} seconds.")
+            return SpannerQueryResult(results, execution_time)
+        except Exception as e:
+            raise RuntimeError(f"Query execution failed: {e}")
+
+    @staticmethod
+    def convert_spanner_json(data):
+        from google.cloud.spanner_v1.data_types import JsonObject
+        json_list = []
+        for item in data:
+            for elements in item:
+                json_entry = {"nodes": [], "edges": []}
+                for element in elements:
+                    element_dict_list = json.loads(element.serialize()) if isinstance(element, JsonObject) else element
+                    for element_dict in element_dict_list:
+                        if element_dict.get('kind') == 'node':
+                            labels = element_dict.get('labels', [])
+                            for label in labels:
+                                node_data = {
+                                    "label": label,
+                                    "identifier": element_dict.get('identifier'),
+                                    "properties": element_dict.get('properties', {})
+                                }
+                                json_entry["nodes"].append(node_data)
+                        elif element_dict.get('kind') == 'edge':
+                            labels = element_dict.get('labels', [])
+                            for label in labels:
+                                edge_data = {
+                                    "label": label,
+                                    "identifier": element_dict.get('identifier'),
+                                    "source": element_dict.get('source_node_identifier'),
+                                    "destination": element_dict.get('destination_node_identifier'),
+                                    "properties": element_dict.get('properties')
```

Comment:
probably too much for this PR, but thinking these would be smoother as:

- if a `Plottable` has node/edge bindings defined, use those
- if not, reuse whatever the graph returns
- take in optional parameters at class constructor or method call time to allow overrides

we're slowly getting better at these, so rough patterns that are emerging


maybe a good Help Wanted / Good First Issue issue to file